File: llm_models.go
package microscope

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"sort"
	"strings"
	"time"
)

type listModelsRequest struct {
	Provider string `json:"provider"`
	APIKey   string `json:"api_key"`
}

type providerModel struct {
	ID        string `json:"id"`
	Label     string `json:"label"`
	CreatedAt string `json:"created_at,omitempty"`
}

func (h *Handler) listLLMModels(w http.ResponseWriter, r *http.Request) {
	r.Body = http.MaxBytesReader(w, r.Body, 8<<10)
	var input listModelsRequest
	if err := json.NewDecoder(r.Body).Decode(&input); err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "invalid JSON body"})
		return
	}
	input.Provider = strings.TrimSpace(strings.ToLower(input.Provider))
	input.APIKey = strings.TrimSpace(input.APIKey)
	if input.Provider == "" || input.APIKey == "" {
		writeJSON(w, http.StatusUnprocessableEntity, map[string]string{"error": "provider and api_key are required"})
		return
	}

	models, err := listProviderModels(r.Context(), input.Provider, input.APIKey)
	if err != nil {
		writeJSON(w, http.StatusBadGateway, map[string]string{"error": err.Error()})
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"models": models})
}

func listProviderModels(ctx context.Context, provider, apiKey string) ([]providerModel, error) {
	switch provider {
	case "openai":
		return listOpenAIModels(ctx, apiKey)
	case "cursor":
		return listCursorModels(ctx, apiKey)
	case "anthropic":
		return listAnthropicModels(ctx, apiKey)
	case "gemini":
		return listGeminiModels(ctx, apiKey)
	default:
		return nil, fmt.Errorf("unsupported provider %q", provider)
	}
}

func providerHTTPClient() *http.Client {
	return &http.Client{Timeout: 25 * time.Second}
}

func providerGET(ctx context.Context, endpoint, apiKey string, headers map[string]string) ([]byte, int, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, endpoint, nil)
	if err != nil {
		return nil, 0, err
	}
	for key, value := range headers {
		req.Header.Set(key, value)
	}
	if apiKey != "" && req.Header.Get("Authorization") == "" && req.Header.Get("x-api-key") == "" && req.Header.Get("x-goog-api-key") == "" {
		req.Header.Set("Authorization", "Bearer "+apiKey)
	}
	res, err := providerHTTPClient().Do(req)
	if err != nil {
		return nil, 0, err
	}
	defer res.Body.Close()
	raw, _ := io.ReadAll(io.LimitReader(res.Body, 4<<20))
	return raw, res.StatusCode, nil
}

func listOpenAIModels(ctx context.Context, apiKey string) ([]providerModel, error) {
	after := ""
	models := make([]providerModel, 0, 64)
	for page := 0; page < 20; page++ {
		endpoint := "https://api.openai.com/v1/models"
		if after != "" {
			endpoint += "?after=" + url.QueryEscape(after)
		}
		raw, status, err := providerGET(ctx, endpoint, apiKey, map[string]string{
			"Authorization": "Bearer " + apiKey,
		})
		if err != nil {
			return nil, err
		}
		if status >= 400 {
			return nil, fmt.Errorf("openai returned %d: %s", status, strings.TrimSpace(string(raw)))
		}
		var parsed struct {
			Data []struct {
				ID      string `json:"id"`
				Created int64  `json:"created"`
			} `json:"data"`
			HasMore bool `json:"has_more"`
		}
		if err := json.Unmarshal(raw, &parsed); err != nil {
			return nil, err
		}
		for _, item := range parsed.Data {
			if !openAIChatModel(item.ID) {
				continue
			}
			created := ""
			if item.Created > 0 {
				created = time.Unix(item.Created, 0).UTC().Format(time.RFC3339)
			}
			models = append(models, providerModel{ID: item.ID, Label: item.ID, CreatedAt: created})
			after = item.ID
		}
		if !parsed.HasMore || len(parsed.Data) == 0 {
			break
		}
	}
	if len(models) == 0 {
		return nil, fmt.Errorf("openai returned no chat-capable models")
	}
	return sortProviderModels(models), nil
}

func openAIChatModel(id string) bool {
	lower := strings.ToLower(id)
	excluded := []string{
		"embedding", "whisper", "tts", "dall-e", "davinci", "babbage", "moderation",
		"realtime", "audio", "transcribe", "search", "sora", "omni-moderation",
		"curie", "ada", "codex", "edit", "similarity", "instruct",
	}
	for _, token := range excluded {
		if strings.Contains(lower, token) {
			return false
		}
	}
	// Include known chat families and any future GPT-family IDs not yet listed.
	if strings.HasPrefix(lower, "gpt-") || strings.HasPrefix(lower, "o1") || strings.HasPrefix(lower, "o3") || strings.HasPrefix(lower, "o4") || strings.HasPrefix(lower, "chatgpt") {
		return true
	}
	return strings.Contains(lower, "gpt") || strings.HasPrefix(lower, "o")
}

func listCursorModels(ctx context.Context, apiKey string) ([]providerModel, error) {
	models, err := listCursorModelsWithAuth(ctx, apiKey, true)
	if err == nil && len(models) > 0 {
		return models, nil
	}
	// Cursor docs also allow Bearer auth.
	models, bearerErr := listCursorModelsWithAuth(ctx, apiKey, false)
	if bearerErr != nil && err != nil {
		return nil, err
	}
	if len(models) == 0 {
		if err != nil {
			return nil, err
		}
		return nil, fmt.Errorf("cursor returned no models")
	}
	return models, nil
}

func listCursorModelsWithAuth(ctx context.Context, apiKey string, basic bool) ([]providerModel, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, "https://api.cursor.com/v1/models", nil)
	if err != nil {
		return nil, err
	}
	if basic {
		req.SetBasicAuth(apiKey, "")
	} else {
		req.Header.Set("Authorization", "Bearer "+apiKey)
	}
	res, err := providerHTTPClient().Do(req)
	if err != nil {
		return nil, err
	}
	defer res.Body.Close()
	raw, _ := io.ReadAll(io.LimitReader(res.Body, 2<<20))
	if res.StatusCode >= 400 {
		return nil, fmt.Errorf("cursor returned %d: %s", res.StatusCode, strings.TrimSpace(string(raw)))
	}
	var parsed struct {
		Items []struct {
			ID          string `json:"id"`
			DisplayName string `json:"displayName"`
			Description string `json:"description"`
		} `json:"items"`
	}
	if err := json.Unmarshal(raw, &parsed); err != nil {
		return nil, err
	}
	models := make([]providerModel, 0, len(parsed.Items))
	for _, item := range parsed.Items {
		if strings.TrimSpace(item.ID) == "" {
			continue
		}
		label := strings.TrimSpace(item.DisplayName)
		if label == "" {
			label = item.ID
		}
		models = append(models, providerModel{ID: item.ID, Label: label})
	}
	return sortProviderModels(models), nil
}

func listAnthropicModels(ctx context.Context, apiKey string) ([]providerModel, error) {
	afterID := ""
	models := make([]providerModel, 0, 32)
	for page := 0; page < 20; page++ {
		query := url.Values{}
		query.Set("limit", "1000")
		if afterID != "" {
			query.Set("after_id", afterID)
		}
		endpoint := "https://api.anthropic.com/v1/models?" + query.Encode()
		raw, status, err := providerGET(ctx, endpoint, apiKey, map[string]string{
			"x-api-key":         apiKey,
			"anthropic-version": "2023-06-01",
		})
		if err != nil {
			return nil, err
		}
		if status >= 400 {
			return nil, fmt.Errorf("anthropic returned %d: %s", status, strings.TrimSpace(string(raw)))
		}
		var parsed struct {
			Data []struct {
				ID          string `json:"id"`
				DisplayName string `json:"display_name"`
				CreatedAt   string `json:"created_at"`
			} `json:"data"`
			HasMore bool   `json:"has_more"`
			LastID  string `json:"last_id"`
		}
		if err := json.Unmarshal(raw, &parsed); err != nil {
			return nil, err
		}
		for _, item := range parsed.Data {
			label := strings.TrimSpace(item.DisplayName)
			if label == "" {
				label = item.ID
			}
			models = append(models, providerModel{ID: item.ID, Label: label, CreatedAt: item.CreatedAt})
		}
		if !parsed.HasMore || parsed.LastID == "" || parsed.LastID == afterID {
			break
		}
		afterID = parsed.LastID
	}
	if len(models) == 0 {
		return nil, fmt.Errorf("anthropic returned no models")
	}
	return sortProviderModels(models), nil
}

func listGeminiModels(ctx context.Context, apiKey string) ([]providerModel, error) {
	pageToken := ""
	models := make([]providerModel, 0, 64)
	seen := make(map[string]struct{})
	for page := 0; page < 20; page++ {
		query := url.Values{}
		query.Set("pageSize", "1000")
		if pageToken != "" {
			query.Set("pageToken", pageToken)
		}
		endpoint := "https://generativelanguage.googleapis.com/v1beta/models?" + query.Encode()
		raw, status, err := providerGET(ctx, endpoint, apiKey, map[string]string{
			"x-goog-api-key": apiKey,
		})
		if err != nil {
			return nil, err
		}
		if status >= 400 {
			return nil, fmt.Errorf("gemini returned %d: %s", status, strings.TrimSpace(string(raw)))
		}
		var parsed struct {
			Models []struct {
				Name                       string   `json:"name"`
				DisplayName                string   `json:"displayName"`
				BaseModelID                string   `json:"baseModelId"`
				SupportedGenerationMethods []string `json:"supportedGenerationMethods"`
			} `json:"models"`
			NextPageToken string `json:"nextPageToken"`
		}
		if err := json.Unmarshal(raw, &parsed); err != nil {
			return nil, err
		}
		for _, item := range parsed.Models {
			if !geminiChatModel(item.Name, item.SupportedGenerationMethods) {
				continue
			}
			id := strings.TrimPrefix(item.Name, "models/")
			if item.BaseModelID != "" {
				id = item.BaseModelID
			}
			if _, ok := seen[id]; ok {
				continue
			}
			seen[id] = struct{}{}
			label := strings.TrimSpace(item.DisplayName)
			if label == "" {
				label = id
			}
			models = append(models, providerModel{ID: id, Label: label})
		}
		if parsed.NextPageToken == "" {
			break
		}
		pageToken = parsed.NextPageToken
	}
	if len(models) == 0 {
		return nil, fmt.Errorf("gemini returned no generative models")
	}
	return sortProviderModels(models), nil
}

func geminiChatModel(name string, methods []string) bool {
	lower := strings.ToLower(name)
	excluded := []string{"embedding", "aqa", "imagen", "veo", "tts", "lyria", "gemma"}
	for _, token := range excluded {
		if strings.Contains(lower, token) {
			return false
		}
	}
	if len(methods) == 0 {
		return strings.Contains(lower, "gemini")
	}
	for _, method := range methods {
		if method == "generateContent" || method == "countTokens" {
			return true
		}
	}
	return false
}

func sortProviderModels(models []providerModel) []providerModel {
	sort.Slice(models, func(i, j int) bool {
		ai, aj := models[i].CreatedAt, models[j].CreatedAt
		if ai != "" && aj != "" && ai != aj {
			return ai > aj
		}
		if ai != "" && aj == "" {
			return true
		}
		if ai == "" && aj != "" {
			return false
		}
		return strings.ToLower(models[i].Label) < strings.ToLower(models[j].Label)
	})
	return models
}

File: adaptor/python/microscope/llm_models.py
from __future__ import annotations

import base64
import json
import urllib.parse
import urllib.request
from typing import Any


def list_provider_models(provider: str, api_key: str) -> list[dict[str, str]]:
    provider = provider.strip().lower()
    if provider == "openai":
        return list_openai_models(api_key)
    if provider == "cursor":
        return list_cursor_models(api_key)
    if provider == "anthropic":
        return list_anthropic_models(api_key)
    if provider == "gemini":
        return list_gemini_models(api_key)
    raise ValueError(f"unsupported provider {provider!r}")


def call_llm(provider: str, model: str, api_key: str, prompt: str) -> str:
    provider = provider.strip().lower()
    if provider in {"openai", "cursor"}:
        endpoint = (
            "https://api.cursor.com/v1/chat/completions"
            if provider == "cursor"
            else "https://api.openai.com/v1/chat/completions"
        )
        payload: dict[str, Any] = {
            "model": model,
            "messages": [
                {"role": "system", "content": "Return only JSON. No markdown."},
                {"role": "user", "content": prompt},
            ],
        }
        if model_supports_custom_temperature(model):
            payload["temperature"] = 0.2
        return post_openai_compatible(endpoint, api_key, payload)
    if provider == "anthropic":
        return call_anthropic(model, api_key, prompt)
    if provider == "gemini":
        return call_gemini(model, api_key, prompt)
    raise ValueError(f"unsupported provider {provider!r}")


def model_supports_custom_temperature(model: str) -> bool:
    lower = model.lower()
    restricted = ("o1", "o3", "o4", "gpt-5", "reasoning", "thinking")
    return not any(token in lower for token in restricted)


def provider_get(endpoint: str, api_key: str, headers: dict[str, str] | None = None) -> tuple[bytes, int]:
    req = urllib.request.Request(endpoint)
    for key, value in (headers or {}).items():
        req.add_header(key, value)
    if api_key and not any(h in req.headers for h in ("Authorization", "x-api-key", "x-goog-api-key")):
        req.add_header("Authorization", f"Bearer {api_key}")
    with urllib.request.urlopen(req, timeout=25) as response:
        return response.read(4 * 1024 * 1024), response.status


def post_openai_compatible(endpoint: str, api_key: str, payload: dict[str, Any]) -> str:
    body = json.dumps(payload).encode("utf-8")
    req = urllib.request.Request(endpoint, data=body, method="POST")
    req.add_header("Content-Type", "application/json")
    req.add_header("Authorization", f"Bearer {api_key}")
    try:
        with urllib.request.urlopen(req, timeout=45) as response:
            raw = response.read(1 * 1024 * 1024)
    except urllib.error.HTTPError as exc:
        raw = exc.read(1 * 1024 * 1024)
        if "temperature" in payload and "temperature" in raw.decode("utf-8", errors="replace"):
            del payload["temperature"]
            return post_openai_compatible(endpoint, api_key, payload)
        raise RuntimeError(f"provider returned {exc.code}: {raw.decode('utf-8', errors='replace').strip()}") from exc
    parsed = json.loads(raw)
    choices = parsed.get("choices", [])
    if not choices:
        raise RuntimeError("provider returned no choices")
    return str(choices[0].get("message", {}).get("content", ""))


def call_anthropic(model: str, api_key: str, prompt: str) -> str:
    payload = {
        "model": model,
        "max_tokens": 1800,
        "messages": [{"role": "user", "content": prompt}],
    }
    body = json.dumps(payload).encode("utf-8")
    req = urllib.request.Request("https://api.anthropic.com/v1/messages", data=body, method="POST")
    req.add_header("Content-Type", "application/json")
    req.add_header("x-api-key", api_key)
    req.add_header("anthropic-version", "2023-06-01")
    with urllib.request.urlopen(req, timeout=45) as response:
        raw = response.read(1 * 1024 * 1024)
    parsed = json.loads(raw)
    content = parsed.get("content", [])
    if not content:
        raise RuntimeError("anthropic returned no content")
    return str(content[0].get("text", ""))


def call_gemini(model: str, api_key: str, prompt: str) -> str:
    endpoint = f"https://generativelanguage.googleapis.com/v1beta/models/{model}:generateContent"
    payload = {
        "contents": [{"parts": [{"text": prompt}]}],
        "generationConfig": {"responseMimeType": "application/json"},
    }
    body = json.dumps(payload).encode("utf-8")
    req = urllib.request.Request(endpoint, data=body, method="POST")
    req.add_header("Content-Type", "application/json")
    req.add_header("x-goog-api-key", api_key)
    with urllib.request.urlopen(req, timeout=45) as response:
        raw = response.read(1 * 1024 * 1024)
    parsed = json.loads(raw)
    candidates = parsed.get("candidates", [])
    if not candidates or not candidates[0].get("content", {}).get("parts"):
        raise RuntimeError("gemini returned no candidates")
    return str(candidates[0]["content"]["parts"][0].get("text", ""))


def list_openai_models(api_key: str) -> list[dict[str, str]]:
    after = ""
    models: list[dict[str, str]] = []
    for _ in range(20):
        endpoint = "https://api.openai.com/v1/models"
        if after:
            endpoint += "?after=" + urllib.parse.quote(after)
        raw, status = provider_get(endpoint, api_key, {"Authorization": f"Bearer {api_key}"})
        if status >= 400:
            raise RuntimeError(f"openai returned {status}: {raw.decode('utf-8', errors='replace').strip()}")
        parsed = json.loads(raw)
        for item in parsed.get("data", []):
            model_id = item.get("id", "")
            if not openai_chat_model(model_id):
                continue
            created = ""
            if item.get("created"):
                from datetime import datetime, timezone

                created = datetime.fromtimestamp(int(item["created"]), tz=timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
            models.append({"id": model_id, "label": model_id, "created_at": created})
            after = model_id
        if not parsed.get("has_more") or not parsed.get("data"):
            break
    if not models:
        raise RuntimeError("openai returned no chat-capable models")
    return sort_provider_models(models)


def openai_chat_model(model_id: str) -> bool:
    lower = model_id.lower()
    excluded = (
        "embedding",
        "whisper",
        "tts",
        "dall-e",
        "davinci",
        "babbage",
        "moderation",
        "realtime",
        "audio",
        "transcribe",
        "search",
        "sora",
        "omni-moderation",
        "curie",
        "ada",
        "codex",
        "edit",
        "similarity",
        "instruct",
    )
    if any(token in lower for token in excluded):
        return False
    if lower.startswith(("gpt-", "o1", "o3", "o4", "chatgpt")):
        return True
    return "gpt" in lower or lower.startswith("o")


def list_cursor_models(api_key: str) -> list[dict[str, str]]:
    try:
        models = list_cursor_models_with_auth(api_key, basic=True)
        if models:
            return models
    except RuntimeError:
        pass
    models = list_cursor_models_with_auth(api_key, basic=False)
    if not models:
        raise RuntimeError("cursor returned no models")
    return models


def list_cursor_models_with_auth(api_key: str, basic: bool) -> list[dict[str, str]]:
    req = urllib.request.Request("https://api.cursor.com/v1/models")
    if basic:
        token = base64.b64encode(f"{api_key}:".encode("utf-8")).decode("ascii")
        req.add_header("Authorization", f"Basic {token}")
    else:
        req.add_header("Authorization", f"Bearer {api_key}")
    with urllib.request.urlopen(req, timeout=25) as response:
        raw = response.read(2 * 1024 * 1024)
    parsed = json.loads(raw)
    models: list[dict[str, str]] = []
    for item in parsed.get("items", []):
        model_id = str(item.get("id", "")).strip()
        if not model_id:
            continue
        label = str(item.get("displayName", "")).strip() or model_id
        models.append({"id": model_id, "label": label})
    return sort_provider_models(models)


def list_anthropic_models(api_key: str) -> list[dict[str, str]]:
    after_id = ""
    models: list[dict[str, str]] = []
    for _ in range(20):
        query = urllib.parse.urlencode({"limit": "1000", **({"after_id": after_id} if after_id else {})})
        endpoint = f"https://api.anthropic.com/v1/models?{query}"
        raw, status = provider_get(
            endpoint,
            api_key,
            {"x-api-key": api_key, "anthropic-version": "2023-06-01"},
        )
        if status >= 400:
            raise RuntimeError(f"anthropic returned {status}: {raw.decode('utf-8', errors='replace').strip()}")
        parsed = json.loads(raw)
        for item in parsed.get("data", []):
            label = str(item.get("display_name", "")).strip() or item.get("id", "")
            models.append(
                {
                    "id": item.get("id", ""),
                    "label": label,
                    "created_at": item.get("created_at", ""),
                }
            )
        if not parsed.get("has_more") or not parsed.get("last_id") or parsed.get("last_id") == after_id:
            break
        after_id = parsed.get("last_id", "")
    if not models:
        raise RuntimeError("anthropic returned no models")
    return sort_provider_models(models)


def list_gemini_models(api_key: str) -> list[dict[str, str]]:
    page_token = ""
    models: list[dict[str, str]] = []
    seen: set[str] = set()
    for _ in range(20):
        query = urllib.parse.urlencode({"pageSize": "1000", **({"pageToken": page_token} if page_token else {})})
        endpoint = f"https://generativelanguage.googleapis.com/v1beta/models?{query}"
        raw, status = provider_get(endpoint, api_key, {"x-goog-api-key": api_key})
        if status >= 400:
            raise RuntimeError(f"gemini returned {status}: {raw.decode('utf-8', errors='replace').strip()}")
        parsed = json.loads(raw)
        for item in parsed.get("models", []):
            if not gemini_chat_model(item.get("name", ""), item.get("supportedGenerationMethods", [])):
                continue
            model_id = str(item.get("name", "")).replace("models/", "")
            if item.get("baseModelId"):
                model_id = item["baseModelId"]
            if model_id in seen:
                continue
            seen.add(model_id)
            label = str(item.get("displayName", "")).strip() or model_id
            models.append({"id": model_id, "label": label})
        page_token = parsed.get("nextPageToken", "")
        if not page_token:
            break
    if not models:
        raise RuntimeError("gemini returned no generative models")
    return sort_provider_models(models)


def gemini_chat_model(name: str, methods: list[str]) -> bool:
    lower = name.lower()
    excluded = ("embedding", "aqa", "imagen", "veo", "tts", "lyria", "gemma")
    if any(token in lower for token in excluded):
        return False
    if not methods:
        return "gemini" in lower
    return any(method in {"generateContent", "countTokens"} for method in methods)


def sort_provider_models(models: list[dict[str, str]]) -> list[dict[str, str]]:
    return sorted(
        models,
        key=lambda item: (
            -(1 if item.get("created_at") else 0),
            item.get("created_at", ""),
            item.get("label", "").lower(),
        ),
    )
